File: solicitation-models/src/main/java/com/solicitation/model/CandidateAttributes.java
package com.solicitation.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Builder;
import lombok.Value;

import javax.validation.constraints.NotNull;
import java.time.Instant;
import java.util.Map;

/**
 * Attributes describing the solicitation opportunity.
 * 
 * These attributes provide properties about the solicitation such as event date,
 * delivery date, channel eligibility, and other business-specific metadata.
 */
@Value
@Builder
public class CandidateAttributes {
    
    /**
     * When the triggering event occurred
     */
    @NotNull
    @JsonProperty("eventDate")
    Instant eventDate;
    
    /**
     * Preferred delivery date (optional)
     */
    @JsonProperty("deliveryDate")
    Instant deliveryDate;
    
    /**
     * Timing window for solicitation (optional)
     */
    @JsonProperty("timingWindow")
    String timingWindow;
    
    /**
     * Order value if applicable (optional)
     */
    @JsonProperty("orderValue")
    Double orderValue;
    
    /**
     * Whether media (images/video) is eligible
     */
    @JsonProperty("mediaEligible")
    Boolean mediaEligible;
    
    /**
     * Channel eligibility flags.
     * Key: channel name (e.g., "email", "in-app", "push")
     * Value: true if eligible for that channel
     */
    @NotNull
    @JsonProperty("channelEligibility")
    Map<String, Boolean> channelEligibility;
}

File: solicitation-models/src/main/kotlin/com/solicitation/model/CandidateAttributes.kt
package com.solicitation.model

import com.fasterxml.jackson.annotation.JsonProperty
import javax.validation.constraints.NotNull
import java.time.Instant

/**
 * Attributes describing the solicitation opportunity.
 *
 * @property eventDate When the triggering event occurred
 * @property deliveryDate Preferred delivery date (optional)
 * @property timingWindow Timing window for solicitation (optional)
 * @property orderValue Order value if applicable (optional)
 * @property mediaEligible Whether media (images/video) is eligible
 * @property channelEligibility Channel eligibility flags
 * @property consumed Whether the candidate has been consumed (delivered)
 * @property consumedAt When the candidate was consumed (optional)
 * @property consumedChannel Channel through which it was consumed (optional)
 */
data class CandidateAttributes(
    @field:NotNull
    @JsonProperty("eventDate")
    val eventDate: Instant,
    
    @JsonProperty("deliveryDate")
    val deliveryDate: Instant? = null,
    
    @JsonProperty("timingWindow")
    val timingWindow: String? = null,
    
    @JsonProperty("orderValue")
    val orderValue: Double? = null,
    
    @JsonProperty("mediaEligible")
    val mediaEligible: Boolean? = null,
    
    @field:NotNull
    @JsonProperty("channelEligibility")
    val channelEligibility: Map<String, Boolean>,
    
    @JsonProperty("consumed")
    val consumed: Boolean = false,
    
    @JsonProperty("consumedAt")
    val consumedAt: Instant? = null,
    
    @JsonProperty("consumedChannel")
    val consumedChannel: String? = null
)
